cmd/xl-zones.go

```diff
@@ -0,0 +1,1085 @@
+	if z.SingleZone() {
+		return z.zones[0].MakeBucketWithLocation(ctx, bucket, location)
+	}
+
+	g := errgroup.WithNErrs(len(z.zones))
+
+	// Create buckets in parallel across all sets.
+	for index := range z.zones {
+		index := index
+		g.Go(func() error {
+			return z.zones[index].MakeBucketWithLocation(ctx, bucket, location)
+		}, index)
+	}
+
+	errs := g.Wait()
+	// Upon even a single write quorum error we undo all previously created buckets.
+	for _, err := range errs {
+		if err != nil {
+			if _, ok := err.(InsufficientWriteQuorum); ok {
+				undoMakeBucketZones(bucket, z.zones, errs)
+			}
+			return err
+		}
+	}
+
+	// Success.
+	return nil
+
+}
+
+func (z *xlZones) GetObjectNInfo(ctx context.Context, bucket, object string, rs *HTTPRangeSpec, h http.Header, lockType LockType, opts ObjectOptions) (gr *GetObjectReader, err error) {
+	if z.SingleZone() {
+		return z.zones[0].GetObjectNInfo(ctx, bucket, object, rs, h, lockType, opts)
+	}
+	for _, zone := range z.zones {
+		gr, err := zone.GetObjectNInfo(ctx, bucket, object, rs, h, lockType, opts)
+		if err != nil {
+			if isErrObjectNotFound(err) {
+				continue
+			}
+			return nil, err
+		}
+		return gr, nil
+	}
+	return nil, ObjectNotFound{Bucket: bucket, Object: object}
+}
+
+func (z *xlZones) GetObject(ctx context.Context, bucket, object string, startOffset int64, length int64, writer io.Writer, etag string, opts ObjectOptions) error {
+	if z.SingleZone() {
+		return z.zones[0].GetObject(ctx, bucket, object, startOffset, length, writer, etag, opts)
+	}
+	for _, zone := range z.zones {
+		if err := zone.GetObject(ctx, bucket, object, startOffset, length, writer, etag, opts); err != nil {
+			if isErrObjectNotFound(err) {
+				continue
+			}
+			return err
+		}
+	}
+	return ObjectNotFound{Bucket: bucket, Object: object}
+}
+
+func (z *xlZones) NewNSLock(ctx context.Context, bucket string, object string) RWLocker {
+	return z.zones[0].NewNSLock(ctx, bucket, object)
+}
+
+func (z *xlZones) GetObjectInfo(ctx context.Context, bucket, object string, opts ObjectOptions) (ObjectInfo, error) {
+	if z.SingleZone() {
+		return z.zones[0].GetObjectInfo(ctx, bucket, object, opts)
+	}
+	for _, zone := range z.zones {
+		objInfo, err := zone.GetObjectInfo(ctx, bucket, object, opts)
+		if err != nil {
+			if isErrObjectNotFound(err) {
+				continue
+			}
+			return objInfo, err
+		}
+		return objInfo, nil
+	}
+	return ObjectInfo{}, ObjectNotFound{Bucket: bucket, Object: object}
+}
+
+// PutObject - writes an object to least used erasure zone.
+func (z *xlZones) PutObject(ctx context.Context, bucket string, object string, data *PutObjReader, opts ObjectOptions) (ObjectInfo, error) {
+	if z.SingleZone() {
+		return z.zones[0].PutObject(ctx, bucket, object, data, opts)
+	}
+	for _, zone := range z.zones {
+		objInfo, err := zone.GetObjectInfo(ctx, bucket, object, opts)
+		if err != nil {
+			if isErrObjectNotFound(err) {
+				continue
+			}
+			return objInfo, err
```

Comment:
So if from another zone we are getting some random error (not related to the actual object itself), we will not proceed to write the object -- is this what we want?

Reply:
Yes correct that is the design @fwessels